src/core/data/grid/gridlayout.h

```diff
@@ -158,8 +159,132 @@ namespace core
 
+            assert(cell_idx == nCells);
+            return std::make_tuple(std::move(x));
+        }
+
+        template<typename Centering, typename StartToEnd, typename CoodsFn>
+        auto gridVectors(DimConst<2>, Centering const& centering, StartToEnd const&& startToEnd,
+                         CoodsFn const& coordsFn, std::size_t const plus) const
+        {
+            auto const [ix0, ix1] = startToEnd(centering, Direction::X);
+            auto const [iy0, iy1] = startToEnd(centering, Direction::Y);
+
+            std::size_t nCells = (ix1 - ix0 + plus) * (iy1 - iy0 + plus);
+            std::vector<double> x(nCells), y(nCells);
+
+            std::size_t cell_idx = 0;
+            for (std::uint32_t ix = ix0; ix < ix1 + plus; ++ix)
+                for (std::uint32_t iy = iy0; iy < iy1 + plus; ++iy)
+                {
+                    auto coord  = coordsFn(ix, iy);
+                    x[cell_idx] = coord[0];
+                    y[cell_idx] = coord[1];
+                    cell_idx++;
+                }
+            assert(cell_idx == nCells);
+            return std::make_tuple(std::move(x), std::move(y));
+        }
+
+        template<typename Centering, typename StartToEnd, typename CoodsFn>
+        auto gridVectors(DimConst<3>, Centering const& centering, StartToEnd const&& startToEnd,
+                         CoodsFn const& coordsFn, std::size_t const plus) const
+        {
+            auto const [ix0, ix1] = startToEnd(centering, Direction::X);
+            auto const [iy0, iy1] = startToEnd(centering, Direction::Y);
+            auto const [iz0, iz1] = startToEnd(centering, Direction::Z);
+
+            std::size_t nCells = (ix1 - ix0 + plus) * (iy1 - iy0 + plus) * (iz1 - iz0 + plus);
+            std::vector<double> x(nCells), y(nCells), z(nCells);
+
+            std::size_t cell_idx = 0;
+            for (std::uint32_t ix = ix0; ix < ix1 + plus; ++ix)
+                for (std::uint32_t iy = iy0; iy < iy1 + plus; ++iy)
+                    for (std::uint32_t iz = iz0; iz < iz1 + plus; ++iz)
+                    {
+                        auto coord  = coordsFn(ix, iy, iz);
+                        x[cell_idx] = coord[0];
+                        y[cell_idx] = coord[1];
+                        z[cell_idx] = coord[2];
+                        cell_idx++;
+                    }
+            assert(cell_idx == nCells);
+            return std::make_tuple(std::move(x), std::move(y), std::move(z));
+        }
+
+        template<typename Centering, typename CoodsFn>
+        auto domainGrids(Centering const& centering, CoodsFn const&& coordsFn,
+                         std::size_t const plus = 0) const
+        {
+            return gridVectors(
+                DimConst<dimension>{}, centering,
+                [&](auto const& centering_, auto const direction) {
+                    return this->physicalStartToEnd(centering_, direction);
+                },
+                coordsFn, plus);
```

Comment:
shouln't coordsFun be std::forwarded?

Reply:
could be, it's a reference so much the same